readme.txt

Evolver, v0.7.1

Date : March 11th, 2003
Author : Damian Murphy
Web : http://minds.may.ie/~murf


============
Introduction
============

Here's some info on the Evolver code.


====================================
1. System Requirements
====================================

I'm unsure what the minimum spec for the code is. The code should compile and
run on any system supporting Sun's JDK 1.3.1 or better. The RAM and disk space
requirements vary with the options used during the code run - for example,
500,000 generations and a population size of 150 used 620Mb of drive space per run.

It's been tested on the following systems, and worked fine:

	Intel Pentium 233, 32MB Ram, Windows 95b
	Intel Pentium 233, 32MB Ram, Red Hat Linux 7
	Intel Pentium IV, 256Mb Ram, Windows NT 4.0 SP6
	Intel Pentium III, 128Mb Ram, Gentoo Linux
	Sun E450 with 4 Sun Sparc CPU's, 2Gb RAM, Sun Solaris 8
	AMD Athlon 850, 128Mb Ram, Windows 2000 SP3
	


====================================
2. Compilation
====================================

Compile all the java source files using 'javac *.java'

You should first choose which genetic operators you want to use for the run, see later for details.


====================================
3. Choosing Operators
====================================

Open the Population.java file with your favourite text editor.
Near the top of this file, you will see something like the following:

// Constants for different implementations
private static final int NO_XCROSS = 1;
private static final int NO_INV = 1;
private static final int NO_MUTP = 1;
private static final int NO_MUTS = 0;
private static final int NO_CLEANUP = 0;

Change these definitions to choose which operators you want to employ.
If they're set to 0 then they are activated, set them to 1 (or some other number) and they will be deactivated for the run. (yes, i know it's reverse logic, i was tired when i coded this!)

The only exception is NO_XCROSS. This is set as follows:

	1 = no crossover
	2 = classic single point crossover
	3 = new variation on single point crossover

Remember to re-compile the code when you've made your changes, or they won't work!


=============================================
4. Running the Code & Command-line parameters
=============================================

To run the code, the following command-line syntax is used:

java Evolver <MUT/INV> <POPSIZE> <CHROMO LENGTH> <GENERATIONS> <PLOTFILE> 

where

<MUT/INV> = Rate of appliction of mutation or inversion operators. Ranges from 0 to 1000.

<POPSIZE> = Size of the population. Any size you like.

<CHROMO LENGTH> = number of cities in the TSP. This parameter is not actually used any more.

<GENERATIONS> = Number of generations to run for.

<PLOTFILE> = Name of file to write the best individual from every 10,000th generation to. This file will be overwritten, so make sure it's got nothing important in it before you start!


Finally, the output is sent to the dos-box screen. Capture this using simple re-direction to a suitable file.

For example, 

java Evolver 2 150 52 500000 run10.plot.txt > run10.dump.txt

sets the mutation rate to 0.02%, pop size at 150, 52 cities, 500,000 generations, saving plot info in run10.plot.txt, and finally capturing the output into the file run10.dump.txt



=============================================
5. Tinkering with the code
=============================================

The code is very specific to the problem I was examining - the effects of operators on genetic algorithms. Poke around with the code, for example the frequency of plot recording (set at every 10,000 generations) is very easy to change - in Evolver.java there's a simple if statement that performs this operation.



=============================================
6. Finally........
=============================================

Have fun!

history.txt
History of changes made to the code for evolver.
Probably missed some changes, can't remember the dates exactly,
most of this was written late at night with a deadline fast approaching!

Damian Murphy (murf),
March 2003.

v0.7.1
Fix initial start city. All functions changed to reflect this.

v0.7
Renamed mutation to inversion. Implemented swap- and point- mutation.

v0.6
GaPlots created.

v0.2.6.1
Old code, forked and brought more up to date to check cyclic behaviour.
Abandoned again when i realise it's just doing a semi-random walk throught the search space.

v0.5.3
No CleanUp. Gives invalid tours a poor fitness.
If a same city is reached again in a tour, add double the distance to the fitness.

v0.5.2
No CleanUp. Gives invalid tours a fitness of MAX_DOUBLE instead.
GA seems to spend all it's time in the wilderness, never getting a valid tour. Chromosomes seem to
homogenise also... why?


v0.5.1
10 Feb 2002
Round-trip TSP with CleanUp. Mutation is an inversion operator.
Single-point crossover.

v0.5
Feb 2002?
non-round trip TSP.

v0.4
Jan/Feb 2002?
Added code to track the ages of individuals.
Doesn't seem to work too well, abandoned again.
Improved all functions again - esp. cleanup & inversion - problems here!
Added code to allow selection of operators & whether to use swing window.
These must be specified by consts at compile time.
will do for now.

v0.3
Jan 2002?
Added no_crossover() function, to transfer parents & children into nex_gen array in correct positions.
This was (and is) done by the crossover() function.
Now we can skip crossover if we wish!
Added code to print best_so_far in numerical order & get length to make it easier to check for errors.
Fixed cleanup, mutation...

v0.2
Jan 20002?
Getting to a solution!
Now doing TSPLIB berlin52 problem.
fitness function changed to reflect this. Now getting distances.
CleanUp introduced.
TSPReader.java created.


v0.1.9
20th Dec 2002
Implemented mutation as the inversion operator - takes a subset of the tour and reverses it.
Parameters for rates, popsize, etc. on command-line.
Created Display.java - a java swing window showing what's happening.

v0.1
Early Dec. 2002. 18th?
First version.
No mutation implemented.
Not doing TSP yet.
Tries to get max( sum genes ) -> e.g. 10 cities, so max is 10*9 = 90, i.e. 9:9:9:9:9:9:9:9:9:9 = chromosome.





Evolver.java
/**
 *      Project 74 : Evolver
 *
 *      See history.txt for versioning information
 *
 *      General test-bed class for the GA Engine
 *
 *      (C)2002 Damian Murphy
 *
 */


import java.io.*;
import java.lang.*;
import java.util.*;


public class Evolver {


    // Global Constants
    public static final String VER = "0.7.1";

    // Set defaults
    static int POP_SIZE;
    static int NUM_GENES;
    static int INV_RATE;     // inversion rate% = INV_RATE/1000
    static int GENERATIONS;
    static int USE_DISPLAY = 1;  // 0 = no window, 1 = use window

    static double best[] = new double[2];
    static double worst[] = new double[2];
    static double[] best_so_far;

    static Display status;

    public static TSPReader TSPInput;

    static long startTime = System.currentTimeMillis();
    static long endTime;

    static GaPlots plotter;
    static String plotfile = "";

    // Class Methods

    private static int parseCommand(String[] args){

            // Retrieve options from the command line
            // and set variables accordingly...
            // Must have certain parameters upon startup, so
            // return -1 to main so we can exit with help given.

            if ( args.length == 0 ){
               return -1;
            }
            else {

              // Now, we need to parse the command args
              //for ( int i=0; i < args.length; i++ ){

              INV_RATE = Integer.parseInt(args[0]);
              POP_SIZE = Integer.parseInt(args[1]);
              NUM_GENES = Integer.parseInt(args[2]);
              GENERATIONS = Integer.parseInt(args[3]);
              plotfile = args[4];

              return 0;
            }
    }



    private static void helpme(){

            // Print out basic command-line help & version info

            System.out.print("\nEvolver, v" + VER + "\n");
            System.out.print("\nUsage:\n");
            System.out.print("\tevolver [options] PARAMETERS\n");
            System.out.print("\nWhere PARAMETERS are :\n");
            System.out.print("  INV_RATE        \t : the inversion rate (0..1000)\n");
            System.out.print("  POPSIZE         \t : population size\n");
            System.out.print("  NUM_GENES       \t : number of genes in each individual\n");
            System.out.print("  GENERATIONS     \t : number of generations to run for\n");
            System.out.print("  PLOTFILE        \t : output file name for plotting info\n\n");


    }

    private static long getEndTime(){

            endTime = System.currentTimeMillis();
            return endTime;

    }

    private static void printRunTime(){

            long runtime, hours, minutes, seconds;

             runtime = endTime - startTime;

             hours = Math.round( (((runtime / 1000)  / 60) / 60) );
             minutes = Math.round( ((runtime / 1000)  / 60) ) % 60 ;
             seconds = Math.round( runtime / 1000) % 60;

             System.out.print( hours + " hours, " + minutes + " minutes, and " + seconds + " seconds.");


    }






    /**
      *     Evolver main method - does all the work!
      *
      **/


    public static void main(String[] args) {

        int errorlevel = 0;
        int sample = 0;


        // Firstly, check & parse the command line
        if ( parseCommand(args) == -1 ){
           helpme();
           System.exit(255);           // Bye-bye!
        }


        TSPInput = new TSPReader();
        NUM_GENES = TSPInput.numCities();

        plotter = new GaPlots( plotfile );
        plotter.init();

        best_so_far = new double[ NUM_GENES + 2 ];

        Population family = new Population (POP_SIZE, NUM_GENES, INV_RATE);

        // Create & show Window
        if ( USE_DISPLAY == 1 ){
           status = new Display( POP_SIZE, NUM_GENES, INV_RATE, GENERATIONS, VER);
           status.show();
        }

        // Dump out time & other stuff.
        System.out.print("Started at UNIX time : " + startTime + "\n\n");
        System.out.print("Evolver v" + VER + "\n");
        System.out.print("Running on " + System.getProperty("os.name") + ", version " + System.getProperty("os.version") + "\nProcessor type : " + System.getProperty("os.arch") + "\n\n");


        for ( int years = 0; years < GENERATIONS; years++){

            System.out.print("Generation : " + family.getGeneration() + "\n\n");

            family.doFitness();

            best = family.getBest();
            worst = family.getWorst();
            best_so_far = family.getBestSoFar();

            System.out.print("Best Pupil : " + (int) best[0]
                              + ", Fitness = " + best[1] /*+ ", Birthday = " + family.getAge( (int) best[0] )*/
                              + "\nWorst Troublemaker : "
                              + (int) worst[0] + ", Fitness = " + worst[1] /*+ ", Birthday = " + family.getAge( (int) worst[0] )*/
                              + "\nAverage Fitness = "
                              + family.getAverage() + "\nBest Fitness so far = " + best_so_far[NUM_GENES] + ", in Generation " + (int) best_so_far[NUM_GENES + 1] + "\n\n");


            // Dump out the top five individuals
            System.out.print("+++ TOP 5 +++\n");
            family.printTopFive();
            System.out.print("=== TOP 5 ===\n\n");

            // Update the window display
            if ( USE_DISPLAY == 1 ){
               status.update( best, worst, family.getGeneration(), family.getAverage(), best_so_far[NUM_GENES], best_so_far[NUM_GENES + 1] );
            }

            // Save plotting info
            if ( sample == 0 ){  // every 10,000 generations, plot a value
               plotter.save( family.getGeneration(), best[1] );
            }


            // Dump out the initial population
            if ( years == 0 ){
            for (int n=0; n < POP_SIZE; n++ ){

                System.out.print("Individual No." + n + " -> ");

                family.printIndividual(n);
                family.printFitness(n);
                System.out.print("\n");

            }
            }

            // Dump out every 'sample' and 'sample+1' population
            /*if ( ( sample == 0 ) || ( sample == 1 ) ){
            for (int n=0; n < POP_SIZE; n++ ){

                System.out.print("Individual No." + n + " -> ");

                family.printIndividual(n);
                family.printFitness(n);
                System.out.print("\n");

            }
            }*/
            // Dump out the final generation
            if ( years == (GENERATIONS - 1) ){
            for (int n=0; n < POP_SIZE; n++ ){

                System.out.print("Individual No." + n + " -> ");

                family.printIndividual(n);
                family.printFitness(n);
                System.out.print("\n");

            }
            }


            family.doSelection();
            System.out.print("\n<->\n\n");
            sample++;
            if ( sample == 10000 ){
              sample = 0;
            }
        }

        // Finished the run.
        // Dump the best ever individual and quit!
        System.out.print("Finished!\n");
        System.out.print("Best Ever Individual was found in generation " + (int) best_so_far[NUM_GENES + 1] + ", with fitness of " + best_so_far[NUM_GENES] + "\n");
        System.out.print("Best ->");

        // Save the best over tour in the plotfile for plotting purposes
        plotter.saveBestInit();       // print headers

         for ( int n=0; n < NUM_GENES; n++ ){
               System.out.print(":" + (int) best_so_far[n] );
               plotter.saveBest( (int) best_so_far[n], TSPInput.getX( (int) best_so_far[n] ), TSPInput.getY( (int) best_so_far[n] ) );
         }

        System.out.print("\n\n");
        System.out.print(" In Numerical Order ->");
        Arrays.sort( best_so_far, 0, NUM_GENES );
        for ( int c=0; c<NUM_GENES; c++){
             System.out.print( (int) best_so_far[c] + ":");
        }


        System.out.print("\n\nRun completed at : " + getEndTime() + "\n");
        System.out.print("Run Time was ");
        printRunTime();
        System.out.print("\n\n");

        // Close plotting file
        plotter.finish();


        System.out.print("\n0 OK, 0:1\n");

        System.exit(0);       // Bye-bye!

    }


}


GaPlots.java
/**
  *     GaPlots.java : Class which dumps plot info into a comma seperated file.
  *     Also dumps out the optimal tour with co-ords for graphing
  *
  */



import java.io.*;
import java.lang.*;
import java.util.*;


public class GaPlots {


       private static String outputFile = new String();
       private static File fd;
       private static FileWriter output;


       public GaPlots( String fileName ){

              outputFile = fileName;

              try{
              fd = new File( outputFile );

              output = new FileWriter( fd );

              }
              catch (IOException ioe)
              {
              System.out.println(ioe.toString());
              }
              catch (SecurityException se)
              {
              System.out.println(se.toString());
              }

       }


       public void init(){

              try{

              output.write("Generated by GaPlots & Evolver v" + Evolver.VER + "\n");
              output.write("Running on " + System.getProperty("os.name") + ", version " + System.getProperty("os.version") + "\nProcessor type : " + System.getProperty("os.arch") + "\n\n\n");
              output.write("Generation,best_fitness,worst_fitness,average_fitness\n");

              }
              catch (IOException ioe)
              {
              System.out.println(ioe.toString());
              }
              catch (SecurityException se)
              {
              System.out.println(se.toString());
              }
       }
       /**
         *    Saves best_fit for each generation, and best_fitness_so_far
         *
         */
       public void save( double best_fit, double best_so_far ){

              try{

              output.write( best_fit + "," + best_so_far + "\n");

              }
              catch (IOException ioe)
              {
              System.out.println(ioe.toString());
              }
              catch (SecurityException se)
              {
              System.out.println(se.toString());
              }
       }

       /**
         *    Saves generation number and best_fit for each generation
         *
         */
       public void save( int generation, double best_fit ){

              try{

              output.write( generation + "," + best_fit + "\n");

              }
              catch (IOException ioe)
              {
              System.out.println(ioe.toString());
              }
              catch (SecurityException se)
              {
              System.out.println(se.toString());
              }
       }

       /**
         *    Saves generation number, best_fit, worst_fit, average_fit
         *
         */

       public void save( int generation, double best_fit, double worst_fit, double average_fit ){

              try{

              output.write( generation + "," + best_fit + "," + worst_fit + "," + average_fit + "\n");

              }
              catch (IOException ioe)
              {
              System.out.println(ioe.toString());
              }
              catch (SecurityException se)
              {
              System.out.println(se.toString());
              }
       }

       public void saveBestInit(){

              try{

              output.write("\n\nBest Ever Tour\n");
              output.write("\ncity,x_coord,y_coord\n");

              }
              catch (IOException ioe)
              {
              System.out.println(ioe.toString());
              }
              catch (SecurityException se)
              {
              System.out.println(se.toString());
              }


       }


       public void saveBest( int city, double x, double y ){

              try{

              output.write( city + "," + x + "," + y + "\n");

              }
              catch (IOException ioe)
              {
              System.out.println(ioe.toString());
              }
              catch (SecurityException se)
              {
              System.out.println(se.toString());
              }


       }



       public void finish(){

              try{

              output.flush();
              output.close();

              }
              catch (IOException ioe)
              {
              System.out.println(ioe.toString());
              }
              catch (SecurityException se)
              {
              System.out.println(se.toString());
              }

       }

}


TSPReader.java
/**
  *     TSPReader : Reads in TSP problems from files
  *
  *
  **/


  // NOTE: This class is temporarily hard coded to fill the array lookuptable with
  // values given below.



import java.io.*;
import java.lang.*;


public class TSPReader {


       // TEMP STATIC ARRAY OF CITIES
       private double[] holder = { 565.0, 575.0, 25.0, 185.0, 345.0, 750.0, 945.0, 685.0, 845.0, 655.0, 880.0, 660.0, 25.0, 230.0, 525.0, 1000.0, 580.0, 1175.0, 650.0, 1130.0, 1605.0, 620.0,  1220.0, 580.0, 1465.0, 200.0, 1530.0, 5.0, 845.0, 680.0, 725.0, 370.0, 145.0, 665.0, 415.0, 635.0, 510.0, 875.0, 560.0, 365.0, 300.0, 465.0, 520.0, 585.0, 480.0, 415.0, 835.0, 625.0, 975.0, 580.0, 1215.0, 245.0, 1320.0, 315.0, 1250.0, 400.0, 660.0, 180.0, 410.0, 250.0, 420.0, 555.0, 575.0, 665.0, 1150.0, 1160.0, 700.0, 580.0, 685.0, 595.0, 685.0, 610.0, 770.0, 610.0, 795.0, 645.0, 720.0, 635.0, 760.0, 650.0,  475.0, 960.0,  95.0, 260.0,  875.0, 920.0, 700.0, 500.0,  555.0, 815.0,  830.0, 485.0, 1170.0, 65.0, 830.0, 610.0, 605.0, 625.0,  595.0, 360.0, 1340.0, 725.0, 1740.0, 245.0 };

       // Globals for this class

       private int NUM_CITIES;

       private double[][] lookUpTable;


       public TSPReader() {

       int x = 0;
       int y = 1;

       NUM_CITIES = 52;

       // Fill up the lookUpTable
       lookUpTable = new double[ NUM_CITIES ][2];

           for ( int n=0; n < NUM_CITIES; n++ ){

               lookUpTable[n][0] = holder[x];
               lookUpTable[n][1] = holder[y];

               x = x + 2;
               y = y + 2;
           }


       }

       /**
         * Returns the number of cities in the TSP
         *
         */
       public int numCities() {

              return NUM_CITIES;
       }

       /**
         * Returns the X Co-ordinate of the city specified by cityID
         *
         */
       public double getX( int cityID ) {

              return lookUpTable[ cityID ][0];
       }

       /**
         * Returns the Y Co-ordinate of the city specified by cityID
         *
         */
       public double getY( int cityID ) {

              return lookUpTable[ cityID ][1];
       }

}







Display.java
/**
  *     Display Class for Printing to Screen...
  *
  *
  **/


import javax.swing.*;
import java.awt.event.*;
import java.awt.*;
import javax.swing.border.*;

public class Display implements ActionListener {


       static JFrame statusBox = new JFrame("Evolver : Status Window");
       static JMenuBar menuBar = new JMenuBar();

       // MenuItems
       private JMenuItem aboutItem;
       private JMenuItem exitItem;

       // Holders
       private static int POP_SIZE;
       private static int NUM_GENES;
       private static int INV_RATE;     // inversion rate% = INV_RATE/1000
       private static int GENERATIONS;
       private static String VERSION;

       // JLabels
       private JLabel generationLabel;
       private JLabel bestLabel;
       private JLabel worstLabel;
       private JLabel averageLabel;
       private JLabel bestsofarLabel;

       public Display(int pop_size, int num_genes, int inv, int lifespan, String VER){

              POP_SIZE = pop_size;
              NUM_GENES = num_genes;
              INV_RATE = inv;
              GENERATIONS = lifespan;
              VERSION = new String( VER );

              statusBox.setBounds(500, 200, 400, 350);
              statusBox.setDefaultCloseOperation( JFrame.EXIT_ON_CLOSE );

              statusBox.setJMenuBar(menuBar);

              // Create Menus
              JMenu fileMenu = new JMenu("File");
              JMenu helpMenu = new JMenu("Help");

              // Add Menu Items
              exitItem = new JMenuItem("Exit");
              aboutItem = new JMenuItem("About");

              fileMenu.add( exitItem );
              helpMenu.add( aboutItem );

              // Add Menus to menuBar
              menuBar.add(fileMenu);
              menuBar.add(helpMenu);

              // Add listeners
              aboutItem.addActionListener( this );
              exitItem.addActionListener( this );



              // Left Frame Content
              Box left = Box.createVerticalBox();
              left.add( Box.createVerticalStrut(5) );
              left.add( new JLabel( "Inv/Mut Rate : " + INV_RATE));
              left.add( new JLabel( "Pop. Size    : " + POP_SIZE));
              left.add( new JLabel( "Gene Length  : " + NUM_GENES));
              left.add( new JLabel( "Lifespan     : " + GENERATIONS + " generations"));


              left.add( Box.createGlue() );


              // Left Frame
              JPanel frameLeft = new JPanel( new BorderLayout() );
              frameLeft.setBorder( new TitledBorder( new EtchedBorder(), "Run Info") );
              frameLeft.add( left, BorderLayout.CENTER );

              // Right Frame Content
              Box right = Box.createVerticalBox();
              right.add( Box.createVerticalStrut(5) );

              // Create labels
              generationLabel = new JLabel( "Generation  : ");
              bestLabel = new JLabel( "Best Individual  : ");
              worstLabel = new JLabel( "Worst Individual : " );
              averageLabel = new JLabel( "Average Fitness  : ");
              bestsofarLabel = new JLabel( "Best So Far : " );

              // Add 'em
              right.add( generationLabel );
              right.add( bestLabel );
              right.add( worstLabel );
              right.add( averageLabel );
              right.add( bestsofarLabel );

              // Right Frame
              JPanel frameRight = new JPanel( new BorderLayout() );
              frameRight.setBorder( new TitledBorder( new EtchedBorder(), "Current Generation") );
              frameRight.add( right, BorderLayout.CENTER );

              // Row to hold left and right
              //Box holder = Box.createHorizontalBox();
              Box holder = Box.createVerticalBox();
              holder.add( frameLeft );
              holder.add( Box.createHorizontalStrut(5) );
              holder.add( frameRight );

              Container content = statusBox.getContentPane();
              BoxLayout box = new BoxLayout( content, BoxLayout.Y_AXIS );

              content.setLayout( new BorderLayout() );
              content.add( holder, BorderLayout.CENTER );

       }



       public void update(double[] best, double[] worst, int generation, double average, double best_fitness, double best_fit_gen){

              // Update evolver status info
              generationLabel.setText( "Generation  : " + generation);
              bestLabel.setText( "Best Individual  : " + (int) best[0] + ", fitness is " + best[1] );
              worstLabel.setText( "Worst Individual : " + (int) worst[0] + ", fitness is " + worst[1] );
              averageLabel.setText( "Average Fitness  : " + average);
              bestsofarLabel.setText( "Best So Far : " + best_fitness + ", in generation " + (int) best_fit_gen );



       }



       static public void show(){

              statusBox.setVisible(true);
       }

       public void actionPerformed( ActionEvent e ){

              if ( e.getSource() == aboutItem){

                 // Show dialog box...
                 AboutDialog aboutEvolver = new AboutDialog( statusBox, "About Evolver");
                 /*JOptionPane.showMessageDialog( this,
                                                "evolver v0.3, (C)2003 D.Murphy, G.Mitchell, NUI Maynooth.",
                                                "About Evolver",
                                                JOptionPane.INFORMATION_MESSAGE);*/
              }
              else
                  if ( e.getSource() == exitItem ){

                     System.exit(254);           // Bye bye!
                  }
       }


       // About dialog box code
       class AboutDialog extends JDialog implements ActionListener{

             public AboutDialog( Frame parent, String title ){

                 super(parent, title, true);

                 if ( parent != null ){
                   Dimension parentSize = parent.getSize();
                   Point p = parent.getLocation();
                   setLocation( p.x+parentSize.width/4, p.y+parentSize.height/4 );
                 }


                 // Create the message pane
                 JPanel messagePane = new JPanel();
                 messagePane.add( new JLabel( "evolver, v" + VERSION ));
                 messagePane.add( new JLabel( "(C)2003 D. Murphy & G. Mitchell," ));
                 messagePane.add( new JLabel( "Dept. Comp. Sci., NUI Maynooth" ));
                 getContentPane().add( messagePane );

                 // Create the ok button and add
                 JPanel buttonPane = new JPanel();
                 JButton okButton = new JButton("OK!");
                 buttonPane.add( okButton );
                 okButton.addActionListener( this );
                 getContentPane().add( buttonPane, BorderLayout.SOUTH );
                 setDefaultCloseOperation( DISPOSE_ON_CLOSE );
                 pack();
                 setVisible( true );

             }

             // OK Button Action
             public void actionPerformed( ActionEvent e ){

                 setVisible( false );
                 dispose();
             }
       }



}
